### ISSUE TITLE
Crash on test after migration to 3.0

### ISSUE THREAD

**Issue by ptitSeb** [wasmerio/wasmer#3638]
Using https://github.com/sedaprotocol/seda-rust on the branch `feat/wasmer3`, doing `cargo test` end prematurely with a segfault.
Something around WASI initialisation seems wrong.

The migration is targetting 3.0 version of Wasmer.

**Comment by ptitSeb** [comment]
Using this patch:
```diff
diff --git a/Cargo.toml b/Cargo.toml
index e0bd197..f3e0200 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -79,5 +79,5 @@ toml = "0.5"
 tracing = "0.1"
 tracing-appender = "0.2"
 tracing-subscriber = { version = "0.3", features = ["env-filter"] }
-wasmer = "3.0"
-wasmer-wasi = "3.0"
+wasmer = { git = "https://github.com/wasmerio/wasmer" }
+wasmer-wasi = { git = "https://github.com/wasmerio/wasmer" }
diff --git a/runtime/core/src/config.rs b/runtime/core/src/config.rs
index 5fb7284..e406be2 100644
--- a/runtime/core/src/config.rs
+++ b/runtime/core/src/config.rs
@@ -1,5 +1,5 @@
 use wasmer::Store;
-use wasmer_wasi::{WasiFunctionEnv, WasiState};
+use wasmer_wasi::{WasiFunctionEnv, WasiEnv};

 #[derive(Clone)]
 pub struct VmConfig {
@@ -17,8 +17,7 @@ pub struct VmConfig {

 impl VmConfig {
     pub fn finalize(self, store: &mut Store) -> WasiFunctionEnv {
-        let mut wasi_state = WasiState::new(&self.program_name);
-        wasi_state.args(&self.args);
-        wasi_state.finalize(store).unwrap()
+        let wasi_state = WasiEnv::builder(&self.program_name);
+        wasi_state.args(&self.args).finalize(store).unwrap()
     }
 }
diff --git a/runtime/core/src/runtime.rs b/runtime/core/src/runtime.rs
index 62bf81e..6070716 100644
--- a/runtime/core/src/runtime.rs
+++ b/runtime/core/src/runtime.rs
@@ -5,7 +5,7 @@ use seda_runtime_adapters::{HostAdapter, InMemory};
 use seda_runtime_sdk::{CallSelfAction, Promise, PromiseAction, PromiseStatus};
 use serde::{Deserialize, Serialize};
 use wasmer::{Instance, Module, Store};
-use wasmer_wasi::{Pipe, WasiState};
+use wasmer_wasi::{Pipe, WasiEnv};

 use super::{imports::create_wasm_imports, PromiseQueue, Result, VmConfig, VmContext};

@@ -84,14 +84,15 @@ impl RunnableRuntime for Runtime {
                         let mut wasm_store = Store::default();

                         // TODO: Use our logging library with stdout/stderr
-                        let mut stdout_pipe = Pipe::new();
-                        let mut stderr_pipe = Pipe::new();
+                        let (stdout_tx, mut stdout_rx) = Pipe::channel();
+                        let (stderr_tx, mut stderr_rx) = Pipe::channel();

-                        let mut wasi_env = WasiState::new(&call_action.function_name)
+                        let mut wasi_env = WasiEnv::builder(&call_action.function_name)
                             .args(call_action.args.clone())
-                            .stdout(Box::new(stdout_pipe.clone()))
-                            .stderr(Box::new(stderr_pipe.clone()))
-                            .finalize(&mut wasm_store)?;
+                            .stdout(Box::new(stdout_tx))
+                            .stderr(Box::new(stderr_tx))
+                            .finalize(&mut wasm_store)
+                            .map_err(|e| e.to_string())?;

                         let current_promise_queue = Arc::new(Mutex::new(promise_queue_mut.clone()));
                         let next_queue = Arc::new(Mutex::new(PromiseQueue::new()));
@@ -110,7 +111,7 @@ impl RunnableRuntime for Runtime {
                         let mut env_mut = vm_context.as_mut(&mut wasm_store);
                         env_mut.memory = Some(wasmer_instance.exports.get_memory("memory")?.clone());

-                        wasi_env.initialize(&mut wasm_store, &wasmer_instance)?;
+                        wasi_env.initialize(&mut wasm_store, wasmer_instance.clone())?;

                         let main_func = wasmer_instance.exports.get_function(&call_action.function_name)?;

@@ -119,12 +120,12 @@ impl RunnableRuntime for Runtime {
                         dbg!(&runtime_result);

                         let mut stdout_buffer = String::new();
-                        stdout_pipe.read_to_string(&mut stdout_buffer)?;
+                        stdout_rx.read_to_string(&mut stdout_buffer)?;
                         dbg!(&stdout_buffer);
                         output.push(stdout_buffer);

                         let mut stderr_buffer = String::new();
-                        stderr_pipe.read_to_string(&mut stderr_buffer)?;
+                        stderr_rx.read_to_string(&mut stderr_buffer)?;
                         dbg!(&stderr_buffer);
                         output.push(stderr_buffer);
```

To migrate to latest version of wasmer (post 3.2.0-alpha.1), and after WASIX migration, the tests are now stable and don't crash anymore. 3 tests are left failling:

```
failures:
    test::runtime_test::test_memory_adapter
    test::runtime_test::test_promise_queue_http_fetch
    test::runtime_test::test_promise_queue_multiple_calls_with_external_traits
```

**Comment by FranklinWaller** [comment]
Hey guys, do we have any idea when this fix will be released? Would love to upgrade :)

**Comment by ptitSeb** [comment]
> Hey guys, do we have any idea when this fix will be released? Would love to upgrade :)

@FranklinWaller I'm not sure I understand the question. Are you asking when the v3.2.0 will be out or are you waiting for some additionnal fixes?

**Comment by syrusakbary** [comment]
We believe this issue is already fixed in Wasmer `3.2.1`, we will close in a few days automatically in case there's no follow up.

Let us know @FranklinWaller if you encounter any other issue from your side

**Comment by FranklinWaller** [comment]
@syrusakbary Yes sorry, got caught up in some other stuff. But this does work now, thank you very much!
